stdlib/random.ml

```diff
@@ -93,6 +93,22 @@ module State = struct
     then invalid_arg "Random.int"
     else intaux s bound
 
+  let rec int63aux s n =
+    let b1 = bits s in
+    let b2 = bits s lsl 30 in
+    let b3 = (bits s land 3) lsl 60 in
```

Comment:
This does the wrong thing on 32-bit systems. Left shifting by `60` is undefined, but comes out as left shifting by `28` on x86. The top four bits of `r` are then the bitwise OR of two uniform random numbers, so end up with a 25% / 75% probability of being 0 or 1, instead of being uniform.

(Arguably this serves you right for calling int63 on a 32-bit system, but it is still pretty weird.)

Reply:
Actually, never mind, I think you can never hit this path on 32-bit systems? You'd need to construct a bound > 0x3FFFFFFF.

Reply:
Yes, exactly, `int63aux` is never called on a 32-bit host.

Reply:
Actually, I was just about to push a correction and ask a question. This does create a problem on JavaScript, and I'm not sure what our opinion on that should be?

Reply:
What problem on JavaScript?

Reply:
It hits exactly this problem, for example with a bound of 0x7FFFFFFF (which is valid in js_of_ocaml)

Reply:
Well, technically it's not exactly this problem - it hits something like it because of representing integers as doubles :)

Reply:
> representing integers as doubles :)

Terrible idea, indeed :-)  The logical "or" of `b1`, `b2`, `b3` will have some low bits always equal to 0.  I would guess the 9 low bits are 0 with probability 1/2, 8 low bits with 3/4, 7 low bits with 7/8, etc.  (I may be off by one.)

Where do we go from here?

Reply:
Actually, I don't understand how ocaml-to-javascript translators compile integer arithmetic.  Is it the case that all integer operations are taken modulo 2^32?

Reply:
I don't think that's quite the problem: Javascript's bitwise or happens on 32-bit integers. (It's roughly: convert double to integer, truncate down to 32 bits, bitwise-or, convert back to double. 32-bit ints can roundtrip through double losslessly, so a sequence of bitwise operations gives the same answer as on actual 32-bit numbers).

Reply:
OK, that was not the problem.  Is it the left shift by 60, then?  Because this can be avoided by factoring:
```
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s land 3 in
    let b = ((b2 lor (b3 lsl 30)) lsl 30) lor b1
```

Reply:
I don't precisely how it works either! In terms of implementation, is having 0x3FFFFFFF < bound <= 0x7FFFFFFF still get 90 random bits but only mix 31 of them disastrous?!

Reply:
As @stedolan wrote, if the left shift by 60 is interpreted as a left shift by 28, you get a bias towards 1 on bits 28, 29, ...  Also, and even with my factoring trick, you can end up with a negative 32-bit integer.

Reply:
Another proposal: first we get an int with all bits random, then we right shift by 1 to make it nonnegative.
```
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s in
    let b = (((b3 lsl 30) lor b2) lsl 30) lor b1 in (* 63 random bits on OCaml 64 bits, 32 random bits on Javascript *)
    let bpos = b lsr 1 in  (* random nonnegative int *)
    ...

Reply:
Does this work for the entire function, therefore?

```ocaml
  let rec int63aux s n =
    let b1 = bits s in
    let b2 = bits s in
    let b3 = bits s in
    (* 62 random bits on 64-bit OCaml; 32 random bits on JavaScript *)
    let b = ((((b3 land 0x30000000) lsl 2) lor b2) lsl 30) lor b1 in
    let bpos = b lsr 1 in (* Random non-negative int *)
    let r =
      (* If 0x3FFFFFFF < n <= 0x7FFFFFFF, take only the low 31 bits,
         for consistent results between 64-bit OCaml and 32-bit JavaScript. *)
      if n < 1 lsl 31 then
        bpos land ((1 lsl 30) + 0x3FFFFFFF)
      else
        bpos
    in
    let v = r mod n in
    if r - v > max_int - n + 1 then int63aux s n else v
```
It's your second proposal, but using the top 2 bits of `b3` as per @damiendoligez's comment.

Reply:
I don't understand the special case `if n < 1 lsl 31`. But if you must special-case `0x3FFFFFFF < n <= 0x7FFFFFFF`, you could just as well combine two `bits s` to get a 31-bit number directly, rather than combine three `bits s` then chop down more bits.

Reply:
Without the special case, you get a different `r mod n` between JavaScript and 64-bit bytecode?

For mixing two calls to `bits` instead of three, that was my previous question - is it better that we _always_ get 90 bits for bounds above `0x3FFFFFFF`, or does it not matter if bounds between `0x3FFFFFFF` and `0x7FFFFFFF` get 60 bits and only bounds above `0x7FFFFFFF` get 90?

Reply:
> > representing integers as doubles :)
> 
> Terrible idea, indeed :-)

(Side-note) You probably know that, but there are no "integer" values in Javascript, just "numbers" (=doubles).  The language still "knows" about integers through its bitwise operators `|`, `&`, `^` which interpret their arguments and result as 32-bit integers.  `(x1 + x2)|0` is thus the standard way to emulate a 32-bit integer semantics.

Reply:
> For mixing two calls to bits instead of three, that was my previous question - is it better that we always get 90 bits for bounds above 0x3FFFFFFF, or does it not matter if bounds between 0x3FFFFFFF and 0x7FFFFFFF get 60 bits and only bounds above 0x7FFFFFFF get 90?

Sorry if I misunderstood the question.  For bounds up to 0x7FFF_FFFF, 60 bits of randomness are plenty to produce a 31-bit integer that will be then reduced to the 0...bound range.  If the bound is just above 0x4000_0000 you may end up iterating more than if you started with a 60-bit random integer, but still it should converge exponentially quickly.